### ISSUE TITLE
pFedHN

### ISSUE THREAD

**Issue by jafermarq** [adap/flower#2028]
# pFedHN

* **Title:** Personalized Federated Learning using Hypernetworks
* **Venue:** ICML 2021
* **Link to paper:** https://proceedings.mlr.press/v139/shamsian21a.html

## Do you want to work on this baseline? 

> 🌻 Check everything about the `Summer of Reproducibility` on [`flower.dev/summer`](https://flower.dev/summer/)

> All available baselines are listed in the [Summer of Reproducibility Dashboard](https://github.com/orgs/adap/projects/6/) and also in the [GitHub Issues with the `summer-of-reproducibility` label](https://github.com/adap/flower/labels/summer-of-reproducibility). The content is the same.

> 📝 It is advised to complete these steps before your start working on your code. But if you can't wait to implement your baseline with Flower (we totally understand it 😄), please ensure you follow the steps on [how to contribute a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline).

What follows are the steps 1 & 2 in the Summer of Reproducibility instructions.

### 1. Join the Summer of Reproducibility program
- [x] Join the [Flower Slack](https://flower.dev/join-slack/) and say _"hi! 👋"_ in the channel `#summer-of-reproducibility`.
- [x] Pick a baseline from our curated list **<---------------------------------------- [you are doing this now]**
### 2. Define the scope of your contribution
- [x] **What are you going to reproduce?** Add a comment to your issue and tell us about your plan regarding this baseline: what experiments from the paper are you reproducing?, for which datasets ? the more details you provide us with the better !
- [x] **Check if you are eligible for a reward**. 
    > As we have to comply with US/EU regulations, we have checked that individual contributors based on these countries or territories are eligible: Australia, Austria, Belgium, Bulgaria, Canada, Croatia, Cyprus, Czech Republic, Denmark, Estonia, Finland, France, Germany, Gibraltar, Greece, Hong Kong SAR China, Hungary, India, Ireland, Italy, Japan, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta, Mexico, Netherlands, New Zealand, Norway, Poland, Portugal, Romania, Singapore, Slovakia, Slovenia, South Korea, Spain, Sweden, Switzerland, Thailand, United Arab emirates, United Kingdom United States.

    If where you are based is not on the list, please send us an email (`summer@flower.dev`) letting us know a bit about yourself (where are you currently based?, are you a university student? do you work at a public institution?). Please tell us the baselines you are interested in implementing (i.e. tell us your GitHub issue if you have crated one). We will reach back to you.
- [x] We will discuss with you about your contribution plan, if it sounds like a substantial enough contribution according to the Summer of Reproducibility rules (check our website [flower.dev/summer](https://flower.dev/summer/)), you'll get the OK from us to start working on your baseline!

## What happens next?

- [x] This item will be moved to the `In Progress` stage by a member of the Flower Team.

- [ ] Follow the [instructions for creating a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline) which will guide you through the process step-by-step.

## Is something wrong or not clear ?
* Ask a question directly in your issue.
* Reach out to us via the [Flower Slack](https://flower.dev/join-slack/) and ask your question in the #summer-of-reproducibility channel
* Check all the details (including FAQ) in the Summer of Reproducibility website: flower.dev/summer

**Comment by lhkhiem28** [comment]
This is a placeholder for pFedHN. I will get back with a plan asap. 
Thanks!

**Comment by lhkhiem28** [comment]
Hi @jafermarq 

This is the contribution plan I'd like to propose: 
- Implement pFedHN using PyTorch and Flower
- Reproduce Table 1: Experiments with pFedHN and pFedHN-PC over 50, 100, and 500 clients on the CIFAR10, and CIFAR100 datasets
- Reproduce Figure 2: Experiments to evaluate pFedHN generalization ability to novel clients using CIFAR10 with 100 clients

I am looking forward to contributing! Happy to discuss this further!

**Comment by jafermarq** [comment]
Hi @lhkhiem28 , the plan you propose looks good! Just a couple of questions:
- For Table 1, it would be great to include one baseline in that table as reference for your `pFedHN` and `pFedHN-PC` rows. How about including `FedPer` (a more basic FL personalisation method)? If you think this would be too much work, including `FedAvg` would be fine. We encourage contributions to include one baseline from the paper (see the first question in our FAQ: [flower.dev/summer](https://flower.dev/summer/)
- Also, could you please check that you are eligible for a reward? See the second point in Step 2. If where you are based is not in the list, feel free to reach out to us by email. If you are associated with a university or a public institution please let us know.

Once we have discussed these two topics I'll ✅ items in Step 1 and Step 2 and you'd be good to start with your contribution!

**Comment by lhkhiem28** [comment]
Hi @jafermarq again, 

- Agree. For Table 1, I can include FedAvg and FedProx as two baselines to compare pFedHN and pFedHN-PC. Because another contributor will work on FedPer, I believe that it’s better to skip it here to save our efforts in this work
- Unfortunately, my country is not on the list. I dropped an email to summer@flower.dev to explain my situation. I am looking forward to hearing from you soon

Thanks!

**Comment by achiverram28** [comment]
Team Information:
We are a team consisting of 2 members Ram Samarth BB(@achiverram28) and Kishan Gurumurthy(@kishan-droid) respectively from India who are federated learning and deep learning enthusiasts . We have extensive enthusiasm in going through,researching and implementing  deep learning models using TensorFlow and are eager to explore novel approaches like Personalized Federated Learning with Hypernetworks.

**Our Perspectives:**

Personalized Federated Learning with Hypernetworks: We believe that this hybrid approach holds great promise for solving the challenges of decentralized data analysis while preserving data privacy. The combination of federated learning and hypernetworks allows us to create personalized models for individual datasets, making the overall system more adaptive and efficient.

Text Data Analysis on UCI ML Repository: Analysing text datasets poses unique challenges.

We would like to reproduce the results with  MNIST, CIFAR-10, and CIFAR-100 datasets using Tensorflow and flower .

**Open-Source Contribution with respect to flower : ** 

We recently contributed to flower in the form of a pull request which got merged successfully.

**How We Will Use This:**

Our team aims to implement the PFH model using TensorFlow and evaluate it on the specified text datasets from the UCI ML Repository.

**Comment by jafermarq** [comment]
Hi @achiverram28 , following our discussion yesterday could you please indicate what is your contribution plan (i.e. what tables/figures you want to reproduce and for which datasets)

**Comment by achiverram28** [comment]
We would like to reproduce the results of Table 1 by working on all the strategies present in it using MNIST , CIFAR-10 and CIFAR100 datasets.

We will implement both the experiments , that is pFedHN and pFedHN-PC

Finally we will also test it on Text Data Analysis on UCI ML Repository

**Comment by jafermarq** [comment]
Sounds good. But let's leave the experiments with the text dataset as the last part to implement (since they are not in the paper). Then this baseline is all yours to implement along with colleagues. It's great to have you on board for the Flower _Summer of Reproducibility_! I have now ✅ all points above for Steps 1 & 2, added you as the _assignee_ of this issue and moved this baseline to _In Progress_ status. If any of your colleagues wants to be included in the _assignee_ list, they should write a comment below (else I GitHub doesn't allow me to select their username).

You can find a guide on how to start with the code by following the link in the _What happens next?_ section above. I encourage you to start soon with your baseline since by the end of September we expect all contributions to be completed. If you have doubts or suggestions please reach out to me or to the other contributors via the Flower Slack workspace.

Really looking forward to seeing `pFedHN` in action using Flower!

**Comment by sachugowda** [comment]
Hi @jafermarq , it's sachin, ram teammate.

**Comment by jafermarq** [comment]
great! @sachugowda , I added you as an _assignee_ as well. Happy to have you onboard !

**Comment by kishan-droid** [comment]
Hey @jafermarq, it's Kishan, Ram's colleague. Could you please add me as an assignee?

**Comment by achiverram28** [comment]
Thanks @jafermarq for giving this opportunity
Excited to bring in the baseline soon

**Comment by jafermarq** [comment]
Hi @achiverram28 , @kishan-droid and @sachugowda!

This is just a gentle reminder that the Flower *Summer of Reproducibility* is ending at the end of the month. With just a little more than 3 weeks to go, we are excited to see quite a few baselines well ahead in the process with [their respective PRs close to ready.](https://github.com/adap/flower/pulls?q=is%3Apr+is%3Aopen+label%3Asummer-of-reproducibility) If your PR is already on the list, great !! Please make sure the PR is linked to this issue (you just need to copy the URL of this issue somewhere in the main message of your PR). Ping me when you'd like me to take a look.

Also, make sure you keep an eye:eyes: on the **#summer-of-reproducibility** channel in the Flower Slack. I’ll announce very soon a new (the third!) round of 1:1 ask-me-anything sessions to help _Summer of Reproducibility_ contributors like yourself to meet the deadline. Please consider booking a time slot if you want to chat with me about your baseline, potential issues you have making your code run, how to open a PR, doubts about what to include in your readme, how to use Hydra configs more effective, etc … all questions are welcome!!

**Comment by WilliamLindskog** [comment]
Closing this as this baseline request has gone stale.
